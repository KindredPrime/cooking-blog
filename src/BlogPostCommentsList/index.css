.BlogPostCommentsList {
  margin-top: 50px;
}