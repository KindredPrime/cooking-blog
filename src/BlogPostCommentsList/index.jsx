import { Component } from 'react';
import PropTypes from 'prop-types';
import CookingContext from '../CookingContext';
import { isOnLastPage } from '../util';
import Comment from '../Comment/index';
import EditComment from '../EditComment/index';
import AddComment from '../AddComment/index';
import TransitionButtons from '../TransitionButtons/index';
import './index.css';

class BlogPostCommentsList extends Component {
  static contextType = CookingContext;

  state = {
    page: 1,
    editingCommentId: null
  };

  static defaultProps = {
    pageLimit: 10,
    comments: []
  };

  // Any errors thrown by this function are caught in AddComment
  handleAdd = (content) => {
    const { user } = this.context;
    return this.props.handleAdd(content, user.id, user.username);
  }

  renderEdit = (id) => {
    this.setState({
      editingCommentId: id
    });
  };

  // Errors are caught and handled by the EditComment component
  handleEditSubmit = (id, content) => {
    return this.props.handleEditSubmit(id, content)
      // Un-render the EditComment
      .then(() => this.setState({
        editingCommentId: null
      }));
  };

  handleEditCancel = () => {
    this.setState({
      editingCommentId: null
    });
  };

  render() {
    const { user } = this.context;
    const { comments, pageLimit, handleDelete } = this.props;
    const { page, editingCommentId } = this.state;

    const displayButtons = comments.length > pageLimit;
    const commentsToRender = comments.slice((page-1) * pageLimit, page * pageLimit);

    return (
      <section className="BlogPostCommentsList">
        <header>
          <h2>Comments</h2>
        </header>

        {user && <AddComment handleSubmit={this.handleAdd} />}

        <ul>
          {commentsToRender.map((comment) => {
            const { id, lastEdited, content, creatorId, creatorUsername, deleted } = comment;

            if (!deleted && editingCommentId && id === editingCommentId) {
              return (
                <EditComment
                  key={`comment-${id}`}
                  id={id}
                  initialContent={content}
                  handleSubmit={this.handleEditSubmit}
                  handleCancel={this.handleEditCancel}
                />
              );
            }

            return (
              <Comment
                key={`comment-${id}`}
                id={id}
                creatorId={creatorId}
                creatorUsername={creatorUsername}
                content={content}
                lastEdited={lastEdited}
                deleted={deleted}
                handleEdit={this.renderEdit}
                handleDelete={handleDelete}
              />
            );
          })}
        </ul>

        {displayButtons
          && <TransitionButtons
            isAtTheBeginning={page === 1}
            isAtTheEnd={isOnLastPage(comments, page, pageLimit)}
            handlePrevClick={() => {
              this.setState({
                page: page - 1
              });
            }}
            handleNextClick={() => {
              this.setState({
                page: page + 1
              });
            }}
          />}
      </section>
    );
  }
}

BlogPostCommentsList.propTypes = {
  comments: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      creatorId: PropTypes.number,
      creatorUsername: PropTypes.string,
      postId: PropTypes.number,
      postTitle: PropTypes.string,
      content: PropTypes.string.isRequired,
      lastEdited: PropTypes.string.isRequired
    })
  ),
  pageLimit: PropTypes.number,
  // handleAdd must return a thenable object
  handleAdd: PropTypes.func.isRequired,
  handleDelete: PropTypes.func.isRequired,
  // handleEditSubmit must return a thenable object
  handleEditSubmit: PropTypes.func.isRequired
};

export default BlogPostCommentsList;