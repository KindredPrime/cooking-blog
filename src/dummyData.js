const dummyPosts = [
  {
    id: 1,
    lastEdited: 'Dec 17 07:37:16 2020',
    title: 'Lessons I Learned from Cooking',
    author: 'FakeUser2',
    content: `Have patience, it can be fun to experiment and explore, having a hard time learning something may only be temporary, bringing an idea of yours to life by creating it from scratch without relying on any instructions can be fun and rewarding, the meals you cook don't have to be fancy or difficult to taste good.`
  },
  {
    id: 2,
    lastEdited: 'Dec 28 18:24:43 2020',
    title: 'How to Save Money When Cooking',
    author: 'User1',
    content: `Buy in bulk, cook in bulk and freeze the leftovers, but store brand products.`
  },
  {
    id: 3,
    lastEdited: 'Dec 1 12:54:31 2020',
    title: '5 Easy Cooking Tips',
    author: 'User3',
    content: `1. When experimenting with seasonings in a recipe, mix them dry in a cup and smell it. If it smells good, then it'll probably taste good.
    2. Add some water when microwaving leftovers
    3. Start small with seasonings
    4. Cook
    5. Blah`
  },
  {
    id: 4,
    lastEdited: 'Jan 5 08:28:28 2021',
    title: 'Cooking Basics for Beginners',
    author: 'User1',
    content: `Go easy on seasonings and salt, if the inside of cooked chicken is pink or its juices aren't clear then it needs to be cooked longer, cooked ground beef needs to be brown in order to be safe to eat`
  },
  {
    id: 5,
    lastEdited: 'Dec 15 11:56:48 2020',
    title: 'Winter Seasoning Guide',
    author: 'User 2',
    content: `Rosemary, sage, thyme, cinnamon, nutmeg`
  },
  {
    id: 6,
    lastEdited: 'Jan 5 14:22:39 2021',
    title: 'Title6',
    author: 'User3',
    content: 'Post 6 content'
  },
  {
    id: 7,
    lastEdited: 'Jan 2 16:22:39 2021',
    title: 'Title7',
    author: 'User3',
    content: 'Post 7 content'
  },
  {
    id: 8,
    lastEdited: 'Jan 6 12:22:39 2021',
    title: 'Title8',
    author: 'User3',
    content: 'Post 8 content'
  },
  {
    id: 9,
    lastEdited: 'Dec 15 11:22:39 2021',
    title: 'Title9',
    author: 'User3',
    content: 'Post 9 content'
  },
  {
    id: 10,
    lastEdited: 'Dec 15 19:22:39 2021',
    title: 'Title10',
    author: 'User3',
    content: 'Post 10 content'
  },
  {
    id: 11,
    lastEdited: 'Jan 4 18:22:39 2021',
    title: 'Title11',
    author: 'User3',
    content: 'Post 11 content'
  }
];

const dummyComments = [
  {
    id: 1,
    lastEdited: new Date('Dec 19 07:37:16 2020'),
    content: 'These are all great lessons!',
    creator: 'User1',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 2,
    lastEdited: new Date('Dec 17 08:37:52 2020'),
    content: `I can't wait to try these seasoning combinations in my next meal!`,
    creator: 'User1',
    postTitle: 'Winter Seasoning Guide'
  },
  {
    id: 3,
    lastEdited: new Date('Jan 5 08:28:28 2021'),
    content: `I wish I knew these when I started learning to cook!`,
    creator: 'User4',
    postTitle: 'Cooking Basics for Beginners'
  },
  {
    id: 4,
    lastEdited: new Date('Dec 31 09:24:43 2020'),
    content: `Great advice!  This'll help me save so much money feeding my three teenagers!`,
    creator: 'User4',
    postTitle: `How to Save Money When Cooking`
  },
  {
    id: 5,
    lastEdited: new Date('Jan 1 10:31:39 2021'),
    content: `What an insightful post`,
    creator: 'FakeUser2',
    postTitle: 'Title7'
  },
  {
    id: 6,
    lastEdited: new Date('Dec 19 08:37:16 2020'),
    content: 'I know, right?',
    creator: 'User3',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 7,
    lastEdited: new Date('Dec 20 10:43:16 2020'),
    content: `It's amazing what you can learn from cooking`,
    creator: 'User4',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 8,
    lastEdited: new Date('Dec 17 20:12:16 2020'),
    content: 'Huh, I never thought about it this way',
    creator: 'User5',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 9,
    lastEdited: new Date('Dec 12 07:07:07 2020'),
    content: 'Comment9',
    creator: 'User6',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 10,
    lastEdited: new Date('Dec 19 15:22:16 2020'),
    content: 'Comment10',
    creator: 'User7',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 11,
    lastEdited: new Date('Dec 15 17:18:19 2020'),
    content: 'Comment11',
    creator: 'User8',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 12,
    lastEdited: new Date('Jan 1 12:13:14 2021'),
    content: 'Comment12',
    creator: 'User5',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 13,
    lastEdited: new Date('Jan 2 13:14:15 2021'),
    content: 'Comment13',
    creator: 'User5',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 14,
    lastEdited: new Date('Jan 5 06:07:08 2021'),
    content: 'Comment14',
    creator: 'User5',
    postTitle: 'Lessons I Learned from Cooking'
  },
  {
    id: 15,
    lastEdited: new Date('Jan 3 20:21:22 2021'),
    content: 'Comment15',
    creator: 'User5',
    postTitle: 'Lessons I Learned from Cooking'
  }
];

const dummyUsers = [
  {
    id: 1,
    username: 'User1',
    password: 'password1',
    email: 'user1@gmail.com'
  },
  {
    id: 2,
    username: 'FakeUser2',
    password: 'password2',
    email: 'fakeuser2@gmail.com'
  },
  {
    id: 3,
    username: 'User3',
    password: 'password3',
    email: 'user3@gmail.com'
  },
  {
    id: 4,
    username: 'User4',
    password: 'password4',
    email: 'user4@gmail.com'
  },
  {
    id: 5,
    username: 'User5',
    password: 'password5',
    email: 'user5@gmail.com'
  },
  {
    id: 6,
    username: 'User6',
    password: 'password6',
    email: 'user6@gmail.com'
  },
  {
    id: 7,
    username: 'User7',
    password: 'password7',
    email: 'user7@gmail.com'
  },
  {
    id: 8,
    username: 'User8',
    password: 'password8',
    email: 'user8@gmail.com'
  }
];

export {
  dummyPosts,
  dummyComments,
  dummyUsers
};