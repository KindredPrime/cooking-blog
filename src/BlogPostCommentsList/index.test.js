import ReactDOM from 'react-dom';
import { render, screen } from '@testing-library/react';
import UserEvent from '@testing-library/user-event';
import { BrowserRouter } from 'react-router-dom';
import BlogPostCommentsList from './index';
import { dummyComments } from '../dummyData';

describe('BlogPostCommentsList Component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <BlogPostCommentsList
        handleAdd={() => {}}
        handleDelete={() => {}}
        handleEditSubmit={() => {}}
      />,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it(`renders the UI as expected`, () => {
    render(
      <BrowserRouter>
        <BlogPostCommentsList
          comments={dummyComments}
          handleAdd={() => {}}
          handleDelete={() => {}}
          handleEditSubmit={() => {}}
        />
      </BrowserRouter>
    );

    expect(document.body).toMatchSnapshot();
  });

  it(`switches to the next group of comments after clicking 'Next'`, () => {
    render(
      <BrowserRouter>
        <BlogPostCommentsList
          comments={dummyComments}
          handleAdd={() => {}}
          handleDelete={() => {}}
          handleEditSubmit={() => {}}
        />
      </BrowserRouter>
    );

    UserEvent.click(screen.getByText('Next'));
    expect(document.body).toMatchSnapshot();
  });

  it(`switches to the previous group of comments after clicking 'Previous'`, () => {
    render(
      <BrowserRouter>
        <BlogPostCommentsList
          comments={dummyComments}
          handleAdd={() => {}}
          handleDelete={() => {}}
          handleEditSubmit={() => {}}
        />
      </BrowserRouter>
    );

    const previous = document.body;

    UserEvent.click(screen.getByText('Next'));
    UserEvent.click(screen.getByText('Previous'));

    expect(document.body).toEqual(previous);
  });
});