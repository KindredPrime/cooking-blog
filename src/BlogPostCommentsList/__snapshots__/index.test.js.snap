// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlogPostCommentsList Component renders the UI as expected 1`] = `
<body>
  <div>
    <section
      class="BlogPostCommentsList"
    >
      <header>
        <h2>
          Comments
        </h2>
      </header>
      <ul>
        <li
          class="Comment"
        >
          <a
            class="Comment__creator"
            href="/users/1"
          >
            User1
          </a>
          <p
            class="Comment__content"
          >
            These are all great lessons!
          </p>
          <p
            class="timestamp"
          >
            Last edited: 
            Dec 19 2020 07:37:16
          </p>
        </li>
        <li
          class="Comment"
        >
          <a
            class="Comment__creator"
            href="/users/1"
          >
            User1
          </a>
          <p
            class="Comment__content"
          >
            I can't wait to try these seasoning combinations in my next meal!
          </p>
          <p
            class="timestamp"
          >
            Last edited: 
            Dec 17 2020 08:37:52
          </p>
        </li>
        <li
          class="Comment"
        >
          <a
            class="Comment__creator"
            href="/users/4"
          >
            User4
          </a>
          <p
            class="Comment__content"
          >
            I wish I knew these when I started learning to cook!
          </p>
          <p
            class="timestamp"
          >
            Last edited: 
            Jan 05 2021 08:28:28
          </p>
        </li>
        <li
          class="Comment"
        >
          <a
            class="Comment__creator"
            href="/users/4"
          >
            User4
          </a>
          <p
            class="Comment__content"
          >
            Great advice!  This'll help me save so much money feeding my three teenagers!
          </p>
          <p
            class="timestamp"
          >
            Last edited: 
            Dec 31 2020 09:24:43
          </p>
        </li>
        <li
          class="Comment"
        >
          <a
            class="Comment__creator"
            href="/users/2"
          >
            FakeUser2
          </a>
          <p
            class="Comment__content"
          >
            What an insightful post
          </p>
          <p
            class="timestamp"
          >
            Last edited: 
            Jan 01 2021 10:31:39
          </p>
        </li>
        <li
          class="Comment"
        >
          <a
            class="Comment__creator"
            href="/users/3"
          >
            User3
          </a>
          <p
            class="Comment__content"
          >
            I know, right?
          </p>
          <p
            class="timestamp"
          >
            Last edited: 
            Dec 19 2020 08:37:16
          </p>
        </li>
        <li
          class="Comment"
        >
          <a
            class="Comment__creator"
            href="/users/4"
          >
            User4
          </a>
          <p
            class="Comment__content"
          >
            It's amazing what you can learn from cooking
          </p>
          <p
            class="timestamp"
          >
            Last edited: 
            Dec 20 2020 10:43:16
          </p>
        </li>
        <li
          class="Comment"
        >
          <a
            class="Comment__creator"
            href="/users/5"
          >
            User5
          </a>
          <p
            class="Comment__content"
          >
            Huh, I never thought about it this way
          </p>
          <p
            class="timestamp"
          >
            Last edited: 
            Dec 17 2020 20:12:16
          </p>
        </li>
        <li
          class="Comment"
        >
          <a
            class="Comment__creator"
            href="/users/6"
          >
            User6
          </a>
          <p
            class="Comment__content"
          >
            Comment9
          </p>
          <p
            class="timestamp"
          >
            Last edited: 
            Dec 12 2020 07:07:07
          </p>
        </li>
        <li
          class="Comment"
        >
          <a
            class="Comment__creator"
            href="/users/7"
          >
            User7
          </a>
          <p
            class="Comment__content"
          >
            Comment10
          </p>
          <p
            class="timestamp"
          >
            Last edited: 
            Dec 19 2020 15:22:16
          </p>
        </li>
      </ul>
      <div
        class="TransitionButtons"
      >
        <button
          class="TransitionButtons__prev"
          disabled=""
          type="button"
        >
          Previous
        </button>
        <button
          class="TransitionButtons__next"
          type="button"
        >
          Next
        </button>
      </div>
    </section>
  </div>
</body>
`;

exports[`BlogPostCommentsList Component switches to the next group of comments after clicking 'Next' 1`] = `
<body>
  <div>
    <section
      class="BlogPostCommentsList"
    >
      <header>
        <h2>
          Comments
        </h2>
      </header>
      <ul>
        <li
          class="Comment"
        >
          <a
            class="Comment__creator"
            href="/users/8"
          >
            User8
          </a>
          <p
            class="Comment__content"
          >
            Comment11
          </p>
          <p
            class="timestamp"
          >
            Last edited: 
            Dec 15 2020 17:18:19
          </p>
        </li>
        <li
          class="Comment"
        >
          <a
            class="Comment__creator"
            href="/users/5"
          >
            User5
          </a>
          <p
            class="Comment__content"
          >
            Comment12
          </p>
          <p
            class="timestamp"
          >
            Last edited: 
            Jan 01 2021 12:13:14
          </p>
        </li>
        <li
          class="Comment"
        >
          <a
            class="Comment__creator"
            href="/users/5"
          >
            User5
          </a>
          <p
            class="Comment__content"
          >
            Comment13
          </p>
          <p
            class="timestamp"
          >
            Last edited: 
            Jan 02 2021 13:14:15
          </p>
        </li>
        <li
          class="Comment"
        >
          <a
            class="Comment__creator"
            href="/users/5"
          >
            User5
          </a>
          <p
            class="Comment__content"
          >
            Comment14
          </p>
          <p
            class="timestamp"
          >
            Last edited: 
            Jan 05 2021 06:07:08
          </p>
        </li>
        <li
          class="Comment"
        >
          <a
            class="Comment__creator"
            href="/users/5"
          >
            User5
          </a>
          <p
            class="Comment__content"
          >
            Comment15
          </p>
          <p
            class="timestamp"
          >
            Last edited: 
            Jan 03 2021 20:21:22
          </p>
        </li>
      </ul>
      <div
        class="TransitionButtons"
      >
        <button
          class="TransitionButtons__prev"
          type="button"
        >
          Previous
        </button>
        <button
          class="TransitionButtons__next"
          disabled=""
          type="button"
        >
          Next
        </button>
      </div>
    </section>
  </div>
</body>
`;
